getting-started/background/index/main.go

```diff
@@ -0,0 +1,165 @@
+
+	pubsubClient, err := pubsub.NewClient(ctx, projectID)
+	if err != nil {
+		return nil, fmt.Errorf("pubsub.NewClient: %v", err)
+	}
+
+	pubsubTopic := pubsubClient.Topic(topicName)
+
+	firestoreClient, err := firestore.NewClient(ctx, projectID)
+	if err != nil {
+		return nil, fmt.Errorf("firestore.NewClient: %v", err)
+	}
+
+	// Template referenced relative to the module/app root.
+	tmpl, err := template.ParseFiles(filepath.Join(templateDir, "index.html"))
+	if err != nil {
+		return nil, fmt.Errorf("template.New: %v", err)
+	}
+
+	return &app{
+		pubsubClient: pubsubClient,
+		pubsubTopic:  pubsubTopic,
+
+		firestoreClient: firestoreClient,
+		tmpl:            tmpl,
+	}, nil
+}
+
+func main() {
+	projectID := os.Getenv("GOOGLE_CLOUD_PROJECT")
+	if projectID == "" {
+		log.Fatalf("GOOGLE_CLOUD_PROJECT must be set")
+	}
+
+	a, err := newApp(projectID, "index")
+	if err != nil {
+		log.Fatalf("newApp: %v", err)
+	}
+
+	http.HandleFunc("/", a.index)
+	http.HandleFunc("/request-translation", a.requestTranslation)
+
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
+	}
+
+	log.Printf("Listening on localhost:%v", port)
+	log.Fatal(http.ListenAndServe(":"+port, nil))
+}
+
+// index lists the current translations.
+func (a *app) index(w http.ResponseWriter, r *http.Request) {
+	docs, err := a.firestoreClient.Collection("translations").Documents(r.Context()).GetAll()
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Error getting translations: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	translations := []background.Translation{}
+	for _, d := range docs {
+		t := background.Translation{}
+		if err := d.DataTo(&t); err != nil {
+			http.Error(w, "Error reading translations", http.StatusInternalServerError)
+			log.Printf("DataTo: %v", err)
+		}
+		translations = append(translations, t)
+	}
+
+	a.tmpl.Execute(w, translations)
+}
+
+// requestTranslation parses the request, validates it, and sends it to Pub/Sub.
+func (a *app) requestTranslation(w http.ResponseWriter, r *http.Request) {
+	r.ParseForm()
+	v := r.PostFormValue("v")
+	if v == "" {
+		log.Printf("Empty value")
+		http.Error(w, "Empty value", http.StatusBadRequest)
+		return
+	}
+	acceptableLanguages := map[string]bool{
+		"de": true,
+		"en": true,
+		"es": true,
+		"fr": true,
+		"ja": true,
+		"sw": true,
+	}
+	lang := r.PostFormValue("lang")
+	if !acceptableLanguages[lang] {
+		log.Printf("Unsupported language: %v", lang)
+		http.Error(w, fmt.Sprintf("Unsupported language: %v", lang), http.StatusBadRequest)
+	}
+
+	log.Printf("Translation requested: %q -> %s", v, lang)
+
+	t := background.Translation{
+		Original: v,
+		Language: lang,
+	}
+	msg, err := json.Marshal(t)
+	if err != nil {
+		log.Printf("json.Marshal: %v", err)
+		http.Error(w, "Error requesting translation", http.StatusInternalServerError)
+		return
+	}
+
+	res := a.pubsubTopic.Publish(r.Context(), &pubsub.Message{Data: msg})
+	if _, err := res.Get(r.Context()); err != nil {
```

Comment:
This races. Wait for Ready before calling Get https://godoc.org/cloud.google.com/go/pubsub#PublishResult.Ready

Reply:
WHAT. Holy cow how did I miss that. Nice :+1: